### PR TITLE
ENH: impl random.noncentral_chisquare

### PR DESCRIPTION
**Pull request by samir-nasibli**
## Description
noncentral_chisquare(df, nonc[, size]) Draw samples from a noncentral chi-square distribution.
* disabled (incorrect check for dpnp.random.noncentral_chisquare functionality):
  * random/tests/test_random.py::TestRandomDist::test_noncentral_chisquare - AssertionError: (external)
  * random/tests/test_randomstate.py::TestRandomDist::test_noncentral_chisquare - AssertionError (external)

## TODO
* array_like of floats for `df` and `nonc ` params
* tests for backend `dpnp_rng_noncentral_chisquare_c` (in other PR, after `backend/tests/test_random.cpp` refactoring PR552 )

###  Refactor TODO
* ``if you use this variable for loops only - use standard way to declare it inside loop boby`` https://github.com/IntelPython/dpnp/pull/543#discussion_r571027739
* "DPNP alloc will take care about errors" - don't check errors for DPNP allocator, see https://github.com/IntelPython/dpnp/pull/543#discussion_r571029614

## Reproduce
```python
>>> numpy.mean(numpy.array(numpy.random.noncentral_chisquare(df, nonc, 10**6))); numpy.mean(numpy.array(dpnp.random.noncentral_chisquare(df, nonc, 10**6)))
21.99867794714586
22.006180859608047
>>> numpy.var(numpy.array(numpy.random.noncentral_chisquare(df, nonc, 10**6))); numpy.var(numpy.array(dpnp.random.noncentral_chisquare(df, nonc, 10**6)))
68.10664345997179
67.90847960630143

```

## Checklist

- [x] Docstrings for all functions
- [ ] Gallery example in ./doc/examples (new features only)
- [x] Unit tests:
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

### CODE REVIEW

**Comment by samir-nasibli** [comment]
@shssf All work on this PR are completed, just waiting some justification for implemented algo for case `df==1`

**Comment by samir-nasibli** [comment]
@shssf @Alexander-Makaryev all works on this PR are completed. Please take a look.

File: dpnp/backend/kernels/dpnp_krnl_random.cpp [hunk 1]
```diff
@@ -533,6 +533,292 @@ void dpnp_rng_negative_binomial_c(void* result, const double a, const double p,
...
+            {
+                size_t* idx = nullptr;
+                _DataType* tmp = nullptr;
+                idx = reinterpret_cast<size_t*>(dpnp_memory_alloc_c(size * sizeof(size_t)));
+                if (idx == nullptr)
+                {
+                    throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+                }
+                for (i = 0; i < size; i++)
+                    idx[i] = i;
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_random.cpp:614]:
```suggestion
                for (int i = 0; i < size; ++i)
                {
                    idx[i] = i;
                }
```

File: dpnp/backend/kernels/dpnp_krnl_random.cpp [hunk 2]
```diff
@@ -533,6 +533,292 @@ void dpnp_rng_negative_binomial_c(void* result, const double a, const double p,
...
+    _DataType* result1 = reinterpret_cast<_DataType*>(result);
+
+    const _DataType d_zero = _DataType(0.0);
+    const _DataType d_one = _DataType(1.0);
+    const _DataType d_two = _DataType(2.0);
+
+    if (dpnp_queue_is_cpu_c())
+    {
+        _DataType shape, loc;
+        size_t i;
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_random.cpp:552]: if you use this variable for loops only - use standard way to declare it inside loop boby

File: dpnp/backend/kernels/dpnp_krnl_random.cpp [hunk 3]
```diff
@@ -533,6 +533,292 @@ void dpnp_rng_negative_binomial_c(void* result, const double a, const double p,
...
+            // or just redesign and return int status:
+            // void dpnp_rng_*_c(..., int& status);
+            // or
+            // int dpnp_rng_*_c(...);
+            if (errcode != VSL_STATUS_OK)
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+            }
+
+            nvec = (double*)mkl_malloc(size * sizeof(double), 64);
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_random.cpp:700]: not sure but it looks like it is better to use DPNP function to allocate the memory

File: dpnp/backend/kernels/dpnp_krnl_random.cpp [hunk 4]
```diff
@@ -533,6 +533,292 @@ void dpnp_rng_negative_binomial_c(void* result, const double a, const double p,
...
+        }
+        else if (df < 1)
+        {
+            /* noncentral_chisquare(df, nonc) ~ G( df/2 + Poisson(nonc/2), 2) */
+            double lambda;
+            int* pvec = nullptr;
+            pvec = (int*)mkl_malloc(size * sizeof(int), 64);
+            if (pvec == nullptr)
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_random.cpp:725]: DPNP alloc will take care about errors

File: dpnp/backend/kernels/dpnp_krnl_random.cpp [hunk 5]
```diff
@@ -533,6 +533,292 @@ void dpnp_rng_negative_binomial_c(void* result, const double a, const double p,
...
+                    // assert(j > i);
+                    if (j <= i)
+                    {
+                        throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+                    }
+                    errcode = vdRngGamma(
+                        VSL_RNG_METHOD_GAMMA_GNORM_ACCURATE, get_rng_stream(), j - i, tmp, shape + cv, d_zero, d_two);
+                    if (errcode != VSL_STATUS_OK)
+                    {
+                        throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_random.cpp:777]: it is better to make one standalone function to check error code

File: dpnp/backend/kernels/dpnp_krnl_random.cpp (hunk 3 above)
Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_random.cpp:700]: @shssf no, it is not good. This part of code for condition when `dpnp_queue` is not `cpu`. So for GPU queue we fallback this function implementation on classic MKL, which works only on CPU. So, I think it is bad scenario using DPNP allocator here.

File: dpnp/backend/kernels/dpnp_krnl_random.cpp (hunk 4 above)
Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_random.cpp:725]: see, https://github.com/IntelPython/dpnp/pull/543#discussion_r571040391

File: dpnp/backend/kernels/dpnp_krnl_random.cpp (hunk 5 above)
Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_random.cpp:777]: Yes, I have the same idea. Let me implement this feature on another PR. I will refactor all related code after.

File: dpnp/backend/kernels/dpnp_krnl_random.cpp (hunk 4 above)
Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_random.cpp:725]: @shssf "DPNP alloc will take care about errors" not here but I have some checks for dpnp alloc on my code, I can generalize this work and make this changes on refactoring for all random functions. I have add this to TODO for refactoring.

**Comment by samir-nasibli** [comment]
@shssf @Alexander-Makaryev I think this feature implementation was completed on this PR. Let me do this refactoring on other PR, see [Descriptions](https://github.com/IntelPython/dpnp/pull/543#issue-563299407), please.
> ###  Refactor TODO
>* ``if you use this variable for loops only - use standard way to declare it inside loop boby``  https://github.com/IntelPython/dpnp/pull/543#discussion_r571027739
> * "DPNP alloc will take care about errors" - don't check errors for DPNP allocator, see https://github.com/IntelPython/dpnp/pull/543#discussion_r571029614

And let me implement this feature on separate PR for all functions and refactor all related code after too. https://github.com/IntelPython/dpnp/pull/543#discussion_r571030198

File: dpnp/backend/kernels/dpnp_krnl_random.cpp (hunk 3 above)
Comment (shssf) [reply dpnp/backend/kernels/dpnp_krnl_random.cpp:700]: You meant it is not SYCL.
hm.. it looks like we need to implement this ability for DPNP allocation functions. To allocate memory with no SYCL queue.

Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_random.cpp:700]: I meant that dpnp own allocator work depends on sycl queue and uses instance of global queue. In this part of code queue will be GPU, but classic mkl uses only cpu mem, so this is not right case using dpnp allocator here.

>hm.. it looks like we need to implement this ability for DPNP allocation functions. To allocate memory with no SYCL queue.

yes, this is good point, we can implement it.

**Review by shssf — APPROVED** [review]

**Review by densmirn — APPROVED** [review]

**Review by Alexander-Makaryev — APPROVED** [review]
🔢

**Pull request merged**